SpatialGDK/Source/SpatialGDKTests/SpatialGDK/LoadBalancing/LayeredLBStrategy/LayeredLBStrategyTest.cpp

```diff
@@ -0,0 +1,412 @@
+	bool FCheckStratIsReady::Update()
+	{
+		Test->TestEqual(
+			FString::Printf(TEXT("Checking strategy for minimum required workers. Actual: %d, Expected: %d"), Strat->IsReady(), Expected),
+			Strat->IsReady(), Expected);
+		return true;
+	}
+
+	DEFINE_LATENT_AUTOMATION_COMMAND_TWO_PARAMETER(FSpawnDefaultPawnAtLocation, FName, Handle,
+		FVector, Location);
+	bool FSpawnDefaultPawnAtLocation::Update()
+	{
+		FActorSpawnParameters SpawnParams;
+		SpawnParams.bNoFail = true;
+		SpawnParams.SpawnCollisionHandlingOverride = ESpawnActorCollisionHandlingMethod::AlwaysSpawn;
+
+		AActor* NewActor = TestWorld->SpawnActor<ADefaultPawn>(Location, FRotator::ZeroRotator, SpawnParams);
+		TestActors.Add(Handle, NewActor);
+
+		return true;
+	}
+
+	DEFINE_LATENT_AUTOMATION_COMMAND_TWO_PARAMETER(FSpawnSpecialPawnAtLocation,
+		FName, Handle, FVector, Location);
+	bool FSpawnSpecialPawnAtLocation::Update()
+	{
+		FActorSpawnParameters SpawnParams;
+		SpawnParams.bNoFail = true;
+		SpawnParams.SpawnCollisionHandlingOverride = ESpawnActorCollisionHandlingMethod::AlwaysSpawn;
+
+		AActor* NewActor = TestWorld->SpawnActor<ASpecialPawn>(Location, FRotator::ZeroRotator, SpawnParams);
+		TestActors.Add(Handle, NewActor);
+
+		return true;
+	}
+
+	DEFINE_LATENT_AUTOMATION_COMMAND_THREE_PARAMETER(FCheckActorsHaveSameAuth,
+		FAutomationTestBase*, Test, FName, FirstActorName, FName, SecondActorName);
+	bool FCheckActorsHaveSameAuth::Update()
+	{
+		VirtualWorkerId FirstActorAuth = Strat->WhoShouldHaveAuthority(*TestActors[FirstActorName]);
+		VirtualWorkerId SecondActorAuth = Strat->WhoShouldHaveAuthority(*TestActors[SecondActorName]);
+
+		Test->TestEqual(
+			FString::Printf(TEXT("Checking actors have the same auth. Actor1: %d, Actor2: %d"), FirstActorAuth, SecondActorAuth),
+			FirstActorAuth, SecondActorAuth);
+		return true;
+	}
+
+	DEFINE_LATENT_AUTOMATION_COMMAND_THREE_PARAMETER(FCheckActorsHaveDifferentAuth, FAutomationTestBase*, Test, FName, FirstActorName, FName, SecondActorName);
+	bool FCheckActorsHaveDifferentAuth::Update()
+	{
+		VirtualWorkerId FirstActorAuth = Strat->WhoShouldHaveAuthority(*TestActors[FirstActorName]);
+		VirtualWorkerId SecondActorAuth = Strat->WhoShouldHaveAuthority(*TestActors[SecondActorName]);
+
+		Test->TestNotEqual(
+			FString::Printf(TEXT("Checking actors have different auth. Actor1: %d, Actor2: %d"), FirstActorAuth, SecondActorAuth),
+			FirstActorAuth, SecondActorAuth);
+
+		return true;
+	}
+
+	DEFINE_LATENT_AUTOMATION_COMMAND_TWO_PARAMETER(FCheckRequiresHandover, FAutomationTestBase*, TEST, bool, Expected);
+	bool FCheckRequiresHandover::Update()
+	{
+		bool Actual = Strat->RequiresHandoverData();
+		TEST->TestEqual(
+			FString::Printf(TEXT("Checking strategy requires handover data Expected: %b Actual: %b"), Expected, Actual),
+			Expected, Actual);
+		return true;
+	}
+
+	DEFINE_LATENT_AUTOMATION_COMMAND_TWO_PARAMETER(FCheckShouldHaveAuthMatchesWhoShouldHaveAuth, FAutomationTestBase*, Test, FName, ActorName);
```

Comment:
This is really checking 2 things.
1) That the LayeredLB queries the same LBStrat for both ShouldHave and WhoShouldHave
2) That the grid based strategy is consistent about ShouldHAve and WhoShouldHave.

Ideally, we'd like to also check that the LBStrat queried is for the correct layer.

I think the only way to do that would be to modify this function to take a virualWorkerId and instead of checking Should and WhoShould, instead check WhoShould against that virtualWorkerId.

Reply:
i have no way of knowing what the correct WhoShould ID is, since the handing out of ids depends on the ordering of strategies in a map.